### DISCUSSION TITLE
Please Explain Inpainting in Depth

### DISCUSSION THREAD

**Discussion by DavidDragonsage** [lllyasviel/Fooocus#2679] [Q&A] [answered]
I like the Inpainting quality in Fooocus, but just when I think I have it all figured out I just get more confused! I've been looking all over this forum for in-depth Inpainting details but not been able to find anything conclusive. So this is what I have been wondering, and sorry for the length of the question, but likely I am not the only one who is confused:

1) I know that Fooocus has a special Inpainting model, but does the base model have any influence? If so, which base model should be used?

2) Do LoRA, styles, Guidance Scale and Image Sharpness have an effect on Inpainting? (I'm pretty sure they all do)

3) There are three Methods of Inpainting. From top to bottom: the "Default", the "Detailer" and the "Modifier". At first it seemed obvious to me what the differences are, but the more Inpainting I do the more I see a confusing overlap between them. For example, Inpainting hands with the Detailer often makes them worse. But using Default or Modifier I often get good results.

What I think is happening is that the Main Prompt (and probably the Negative Prompt) are always active when you are Inpainting, no matter which of the three Methods you are using.

I think this is why I sometimes get a tiny version of the Main Prompt - like a character's head - when I'm using the Modifier. So it seems to me I should just erase the Main Prompt when I am doing a Modification/Replacement (which makes me wonder why the Main Prompt should be permitted to have any influence at all in the Modification mode because it just causes trouble).

I think that the "Default" method is controlled by both the Main Prompt and by the logic of the generator in trying to create a blend with the rest of the image, i.e. what makes sense according to the image content. And likely it is also effected by LoRA, styles, etc. as mentioned in (2) above.

Lastly, perhaps the Detailer may be less influenced by the Main Prompt. I have never seen it try to put a character's head or add any strange stuff other than ornamentation when I've been Inpainting a face for example.

**Comment by mashb1t** [comment, marked as answer]
> 1. I know that Fooocus has a special Inpainting model, but does the base model have any influence? If so, which base model should be used?

The base model (+ refiner + LoRAs) you chose + the inpaint model are used, which is also why inpaint in general uses more system resources (min. 2 models). See https://github.com/lllyasviel/Fooocus/blob/main/modules/async_worker.py#L429 and https://github.com/lllyasviel/Fooocus/blob/main/modules/async_worker.py#L682, patch in https://github.com/lllyasviel/Fooocus/blob/main/modules/inpaint_worker.py#L194

> 2. Do LoRA, styles, Guidance Scale and Image Sharpness have an effect on Inpainting? (I'm pretty sure they all do)

Yes, all of them are taken into account.

> 3. There are three Methods of Inpainting. From top to bottom: the "Default", the "Detailer" and the "Modifier". At first it seemed obvious to me what the differences are, but the more Inpainting I do the more I see a confusing overlap between them. For example, Inpainting hands with the Detailer often makes them worse. But using Default or Modifier I often get good results.

You can see the details what the three mode do when opening advanced checkbox > advanced tab > developer debug mode checkbox > inpaint tab or in https://github.com/lllyasviel/Fooocus/blob/main/webui.py#L591-L616. 

The methods modify:
1. inpaint_additional_prompt
2. outpaint_selections
3. example_inpaint_prompts
4. inpaint_disable_initial_latent
5. inpaint_engine
6. inpaint_strength
7. inpaint_respective_field

Modes:

1. `Default`
Similar to Modify Content, but taking ~60% context (non-masked) into account
2. `Improve Detail`
No inpainting engine, just 0.5 inpaint_strength, only masked
3. `Modify Content`
No context at all, only masked, full inpaint_strength, more or less completely disconnected generation

All modes take your prompts (positive, negative) into account, additional inpaint prompt is supported by `Improve Detail` and `Modify Content`.

Hope this helps in understanding inpaint in Fooocus better. Is there still something unclear?

**Reply by DavidDragonsage** [reply]
Wow! I don't know where you find the time to reply in this depth but thank you for an incredibly detailed response!

As well as you providing an excellent summary, I really appreciated looking at the code too. It looks very clean to me. All this will definitely help me in deciding which mode to use and why they work the way they do.

So Improve Detail just uses the base model + refiner + LoRAs? It should operate more quickly but it also means that the choice of base model is quite critical. I have a fairly good understanding of the capabilities of the base models I am using - their strengths and weaknesses - so I will use that information for best effect.

For example I use the Faetastic base model a lot for my fantasy work, but it would make sense to switch to Juggernaut 8 or Juggernaut 9 when doing facial Inpainting because they both excel at portraits. There is much food for thought here - thanks again! 🙂

**Reply by ILikeToasters** [reply]
One thing to keep in mind with the improve detail is that it will not often fix bad stuff. For example if you try to improve detail on a hand with 6 fingers it will just be a higher quality six finger hand most likely. But it does well if the basic structure is good.

**Reply by DavidDragonsage** [reply]
Yes, that's what I found too - six perfect fingers!
